### DISCUSSION TITLE
Unable to produce a valid signature

### DISCUSSION THREAD

**Discussion by luisgomesopc** [MatthiasValvekens/pyHanko#349] [Q&A]
Hello @MatthiasValvekens,

First of all, I want to thank you and congratulate you for all the work and dedication you've put into this project!

I'm currently trying to use an External Signer to digitally sign a pdf document, but upon validating said signature using the method 'validate_pdf_signature', I haven't been able to get another result other than 'The signature is judged INVALID'.
The validation report always specifies that the signature is cryptographically unsound and that the signer's certificate is untrusted.

Both the certificate chain and the hash of the file that I'm using to sign the documents are all provided by an external entity, so unless I'm using them incorrectly, I'm not really understanding what I'm doing wrong.

Another thing that I've noticed, is that when I open the signed file and read it's contents (using 'with open(...)'), I can see 2 Signature Objects, both with the same name that I specified when I created the SigFieldSpec.

Here is the method I use to generate the signature container:

```python
@staticmethod
 def generate_signature_container(pdf_file):
     pdf_bytesio = io.BytesIO(pdf_file)
     sig_field_spec = SigFieldSpec(sig_field_name='Signature1', box=(10, 10, 150, 50), on_page=-1)
     pdf_writer = IncrementalPdfFileWriter(pdf_bytesio)
     append_signature_field(pdf_writer, sig_field_spec)
     pdf_writer.write_in_place()
     return pdf_bytesio.getvalue()
``` 

And here is the code that I'm using to create a PdfSigner and sign the document

```python
        pdf_signer = signers.PdfSigner(
            signers.PdfSignatureMetadata(
                field_name='Signature1',
                md_algorithm='sha256',
            ),
            signer=signers.ExternalSigner(
                signing_cert=signing_cert_pem,
                # signing_cert, issuer_cert and root_cert are the certificates returned by the third party in string format
                cert_registry=SimpleCertificateStore.from_certs(
                    [signing_cert_pem, issuer_cert_pem, root_cert_pem]
                ),
                # signatures['signatures'][0] is the signed hash returned by the third party in string format
                signature_value=base64.b64decode(signatures['signatures'][0]),
                embed_roots=True,
            ),
        )
```
Once again, appreciate your efforts.
Thanks in advance.

**Comment by MatthiasValvekens** [comment]
Hi, this indeed happens quite often ;).

The main issue with your code is that at the point you're creating the `PdfSigner`, you _definitely_ don't have a valid signature yet. Whatever digest you fed to the external signing service, it's not going to work. The reason is that the digest you need to compute is taken over all data in the PDF document except for the signature container itself. Incidentally, this container is also not the same thing as the signature _field_.

Long story short: you need to let pyHanko do the PDF-related heavy lifting. You only need to supply the plumbing to talk to that external service. There are however multiple API conventions for such external services, and they're mutually incompatible & often poorly documented. Without specific documentation about the service that you're using, it's hard to tell you exactly what to do, but there are various resources available (search "interrupted signing" on this forum, or take a look here for some common usage patterns: https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing) 

PS: from the snippet you shared, it looks like the service you're using might actually be implementing the Cloud Signature Consortium API. In that case, you might even be able to work with `CSCSigner` instead of `ExternalSigner`: https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.signers.html?highlight=CSCSigner#cscsigner-overview. If you do that, you don't have to worry about all the interrupted signing complications, unless you need the flexibility to do your own performance optimisations.
